File: solutions/scala/sieve/1/src/main/scala/Sieve.scala
object Sieve {
  def primes(n: Int): Seq[Int] = sieve(Stream.from(2)) takeWhile { _ <= n }
  def sieve(s: Stream[Int]): Stream[Int] = s.head #:: sieve(s.tail filter { _ % s.head != 0} )
}

File: solutions/erlang/sieve/1/src/sieve.erl
-module(sieve).

-export([primes/1]).


primes(Limit) -> sieve(lists:seq(2, Limit), trunc(math:sqrt(Limit))).

sieve([Prime | T], Limit) when Prime =< Limit ->
    [Prime | sieve(lists:filter(fun(Elem) -> Elem rem Prime > 0 end, T), Limit)];
sieve(Lst, _) -> Lst.
